### PR TITLE
More thread-safety changes

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Could the arena case not be handled by wrapping arenas in a mutex instead of hard-coding the mutex into the arena types? This way one could decide on a case-by-case basis if an arena needs to be synchronized or not.

**Comment by Zoxc** [comment]
No, arenas and mutexes do not compose. We could have a thread-safe and thread-unsafe variants of the arenas however. That would probably reduce the amount of changed lines too.

**Comment by michaelwoerister** [comment]
Can you elaborate on that? Do you mean that one can't use a thread-unsafe arena where a thread-safe arena is expect?

**Comment by Zoxc** [comment]
If you use a `Mutex<Arena>` the lifetimes of the allocated objects will be restricted to the mutex guard.

**Comment by Zoxc** [comment]
I added thread-safe variant of the arenas instead of modifying the existing types.

**Comment by michaelwoerister** [comment]
> I added thread-safe variant of the arenas instead of modifying the existing types.

Thanks, I think that's an improvement.

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -1554,7 +1554,7 @@ impl<'gcx: 'tcx, 'tcx> GlobalCtxt<'gcx> {
     /// Call the closure with a local `TyCtxt` using the given arena.
     pub fn enter_local<F, R>(
         &self,
-        arena: &'tcx DroplessArena,
+        arena: &'tcx SyncDroplessArena,
```
Comment (michaelwoerister) [inline src/librustc/ty/context.rs]: Could the local arenas in theory be non-synchronized?

Comment (Zoxc) [reply src/librustc/ty/context.rs]: All the arenas could be thread-local if we find a way to deal with lifting of types.

**Pull request merged**